### PR TITLE
Mkdocs initial POC with code optioning

### PR DESCRIPTION
**Pull request by JasonBellONS**
# Description

Mkdocs POC implemented along with code optioning to prevent it from being visible on the site.

Included in this PR is an initial structure of how mkdocs would be built into the website, done from the perspective of files being obtained from another repo, full integration will include mkdocs being built from this repo instead but is out of scope for this PR

This PR includes a number of pages providing api and method specifications for Totals and Components, as well as cell key perturbation, as well as linking these pages from the methods page

Additionally, code optioning has been added in to prevent these pages from being visible on the website, instead users will be redirected to a work in progress page (404'ing the page was not an option as it would trigger an error from frozen flask and fail the pipeline). Any site navigation will not include the api documentation and any links that would take a user to the api documentation has also been optioned out unless it is turned back on via the config files

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Technical Enhancements

# Checklist:

If any of these are not completed, please explain why in the notes.

## **Definition of Done**
**Code and merges**

- [x] Code to be commented on where applicable 
- [x] Documentation updated where required 
- [x] Have considered non-functional requirements such as Security, Performance, Scalability, and Fault Tolerance
- [x] I have linted the code

**Testing**

- [x] All levels of acceptance test are passing (automated, integration, manual, accessibility, etc.)
- [x] I have run the behave command to check the selenium behaviour tests pass locally
- [x] Acceptance criteria met

**Other Checks**
- [x] I have performed a self-review of my own code
- [x] I have checked for spelling errors
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] My changes don't break anything unexpected
- [x] I have checked and updated the security.txt file where required
- [x] Up to date with the main branch
- [x] For a change that needs to be deployed in a release I have a commit with the fix:, feat: or BREAKING CHANGE:
tags and will copy this commit message to the squash
- [ ] For a change that does not affect the deployed code I have added a commit with ci:, docs: or test:
depending upon which area of the solution was affected

Note when merging, squash commit must begin with one of the following tags:
"BREAKING CHANGE", "feat", "fix", "ci", "docs", "test"

Usage

BREAKING CHANGE: any change that breaks current functionality and may cause compatibility errors to users upgrading
    E.G BREAKING CHANGE: refactored method x to take variable y as input (changed from variable z)

feat: A new feature
    E.G feat: adding method x to future methods table

fix: A bug fix
    E.G fix: amending page for method x to show correct release version

ci: Changes to our CI configuration files and scripts (Concourse)

docs: Documentation only changes

test: Adding missing tests or correcting existing tests

### CODE REVIEW

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://dx4x6oa31y3qw.cloudfront.net/
cloudfront_id=E2SHB7A4J3FTNH

**Review by gibbardsteve — COMMENTED** [review]
Can you update the PR description to give me a bit more of an overview on the changes you have made please?  It'll be easier to review with some background.

File: sml_builder/templates/method.html [hunk 1]
```diff
@@ -109,7 +109,12 @@ <h2>GitHub resources</h2>
     "linkText":"Go to "~ page.title ~" user documentation on GitHub"
   })
 }}
-
 {% endif %}
-
+<br>
+{% if mkdocs_active == True %}
+    {% if page.api_reference %}
+    {%- set api_url = url_for(".api_guidances", category="api-reference", subcategory="totals-and-components") -%}
```
Comment (gibbardsteve) [inline sml_builder/templates/method.html:116]: Should be indented

File: terraform/cloudfront.tf [hunk 2]
```diff
@@ -108,7 +108,7 @@ resource "aws_cloudfront_response_headers_policy" "noindex" {
     }
     content_security_policy {
       override                = true
-      content_security_policy = "script-src 'self' 'sha256-AN/TTkrWvH4W7oCH0w79aph67vAli7rskVQ+0LFPnD4=' 'sha256-FM5wgoZgD+IUY/B3phTcQwLPjMdoyegYYa9Lfsx8a0Q=' 'sha256-Khdmw7GCwPLFjfMGBMpTNcdMf4c+bxQbA3kB831L6KM=' 'sha256-wVoJJ46dUdegqo0mzHo83JJUyjS8RdEfN7I1OlrdMOU=' 'sha256-Lp8Q2eLwjHw/eWeZ5jdkGBt5wP7gIjrZXTt4JuVpYQs=' 'sha256-EMdG31ovSprR0tF3JVid3tnd4Db/rb+0c27xK7EAVts=' https://cdn.ons.gov.uk https://www.googletagmanager.com"
+      content_security_policy = "script-src 'self' 'sha256-AN/TTkrWvH4W7oCH0w79aph67vAli7rskVQ+0LFPnD4=' 'sha256-FM5wgoZgD+IUY/B3phTcQwLPjMdoyegYYa9Lfsx8a0Q=' 'sha256-FM5wgoZgD+IUY/B3phTcQwLPjMdoyegYYa9Lfsx8a0Q=' 'sha256-QTkwzjnbU+Qv8p+Xz6cexltGK9UI2ShI34bF9k9DpXs=' 'sha256-EMdG31ovSprR0tF3JVid3tnd4Db/rb+0c27xK7EAVts=' 'sha256-Khdmw7GCwPLFjfMGBMpTNcdMf4c+bxQbA3kB831L6KM=' 'sha256-wVoJJ46dUdegqo0mzHo83JJUyjS8RdEfN7I1OlrdMOU=' 'sha256-Lp8Q2eLwjHw/eWeZ5jdkGBt5wP7gIjrZXTt4JuVpYQs=' 'sha256-EMdG31ovSprR0tF3JVid3tnd4Db/rb+0c27xK7EAVts=' https://cdn.ons.gov.uk https://www.googletagmanager.com https://cdnjs.cloudflare.com"
```
Comment (gibbardsteve) [inline terraform/cloudfront.tf:111]: what is the cdnjs.cloudflare.com policy for?

File: config/feature.json [hunk 3]
```diff
@@ -1,5 +1,42 @@
 {
     "features": {
-        "FEATURE_1": false
+        "FEATURE_1": false,
+        "MKDOCS": {
```
Comment (gibbardsteve) [inline config/feature.json:4]: mkdocs is the tool, could we name the feature docs_integration ?  We will be having some discussions with business about the features and I am trying to find 'user friendly' terminology to help that conversation. If we can use the same terminology in the code it'll be helpful (and also means if we had to swap out mkdocs for a different doc builder, unlikely, but possible then we don't have to update the references).

File: sml_builder/__init__.py [hunk 4]
```diff
@@ -40,3 +44,20 @@ def string_to_paragraph(value):
     """Wraps passed string in <p> tags and converts newlines to <p> pairs"""
     body = escape(value).replace("\n\n", Markup("</p><p>"))
     return Markup(f"<p>{body}</p>")
+
+
+@app.context_processor
```
[the remark this answers is not in the document]
Comment (JasonBellONS) [reply sml_builder/__init__.py:49]: yep thats right a context processor makes the dictionary returned here available within the jinja templates, ill update this to explain

File: sml_builder/api_reference.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    contents_helper,
+    get_feature_config,
+)
+
+
+@app.route("/api_reference/index")
+def api_reference(category=None):
+    mkdocs = get_feature_config("MKDOCS")
+    if mkdocs["ACTIVE"] is False:
+        return Response("Feature is a work in Progress")
```
[the remark this answers is not in the document]
Comment (JasonBellONS) [reply sml_builder/api_reference.py:18]: This currently just drops a blank html page with the string above on it, similar to the security.txt files, its only reachable to a user if they specifically type the full url into a browser which they'd only really know if they were poking around on this repo, happy to change it to be more styled like the 404 page

File: sml_builder/api_reference.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    return render_template(
+        "api-index.html", help_categories=categories, selected_category=category
+    )
+
+
+@app.route("/api_reference/<category>/index")
+@app.route("/api_reference/<category>/<sub_category>")
+def api_guidances(category, sub_category=None):
+    mkdocs = get_feature_config("MKDOCS")
+    if mkdocs["ACTIVE"] is False:
```
[the remark this answers is not in the document]
Comment (JasonBellONS) [reply sml_builder/api_reference.py:36]: done, as above

File: sml_builder/api_reference.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+                    "anchors": [
+                        {
+                            "title": sub_category["label"],
+                            "url": url_for(
+                                "api_guidances",
+                                category=category["name"],
+                                sub_category=sub_category["name"],
+                            ),
+                        }
+                        for sub_category in category["subcategories"]
```
[the remark this answers is not in the document]
Comment (JasonBellONS) [reply sml_builder/api_reference.py:84]: pretty much, its rebuilding the .json file into a format the ons design system can use and while it does converts the url fields into actual URL objects that flask can use, will add a comment

File: sml_builder/templates/api-docs/api-reference/create_perturbed_table/create_perturbed_table.html
[the remark this answers is not in the document]
Comment (JasonBellONS) [reply sml_builder/templates/api-docs/api-reference/create_perturbed_table/create_perturbed_table.html:1]: yep these files are the static html generated by mkdocs, can potentially look at moving them into the .gitignore once we move mkdocs into this repo and we can get a better process for this running

File: sml_builder/templates/api-docs/index.html [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+                
+
+            </div>
+
+        <div class="container">
+            <div class="row">
+                    <div class="col-md-9" role="main"><section class="p-md-5 section-content">
+    <article>
+        <p><h1 id="sml-small-api-reference">SML-Small API Reference</h1>
+<p>testy testy banana</p></p>
```
[the remark this answers is not in the document]
Comment (JasonBellONS) [reply sml_builder/templates/api-docs/index.html:15]: Forgot I had to make a main page :) will amend

File: terraform/cloudfront.tf (hunk 2 above)
Comment (JasonBellONS) [reply terraform/cloudfront.tf:111]: this is what powers the mathjax in the method spec markdown files

File: config/feature.json (hunk 3 above)
Comment (JasonBellONS) [reply config/feature.json:4]: done

File: pyproject.toml [hunk 9]
```diff
@@ -9,7 +9,9 @@ readme = "README.md"
 python = "^3.10"
 frozen-flask = "^0.18"
 jsonnet = "^0.20.0"
-markdown = "^3.4.4"
+markdown = "*"
+python-markdown-mathjax = "1.0.4"
+python-markdown-math="*"
```
[the remark this answers is not in the document]
Comment (JasonBellONS) [reply pyproject.toml:14]: done ^0.8

File: sml_builder/__init__.py [hunk 10]
```diff
@@ -26,7 +27,10 @@
 
 @app.route("/")
 def index():
-    return render_template("index.html")
+    return render_template(
+        "index.html",
+        feature_1_enabled=sml_builder.utils.get_feature_config("FEATURE_1"),
```
[the remark this answers is not in the document]
Comment (JasonBellONS) [reply sml_builder/__init__.py:32]: not 100% sure what you mean but black is very insistent on that being the format

File: config/feature.json [hunk 11]
```diff
@@ -1,5 +1,42 @@
 {
     "features": {
-        "FEATURE_1": false
+        "FEATURE_1": false,
```
[the remark this answers is not in the document]
Comment (JasonBellONS) [reply config/feature.json:3]: amended to lower case, ive also renamed the lists to feature_active_navigation and feature_disabled_navigation to make it more clear what these two lists do

File: config/feature.json [hunk 12]
```diff
@@ -1,5 +1,42 @@
...
+        "MKDOCS": {
+            "ACTIVE": false,
```
[the remark this answers is not in the document]
Comment (JasonBellONS) [reply config/feature.json:5]: happy to swap it to enabled, thats correct the top bar navigation is powered via this config now, when the feature is disabled we get the navigation excluding the api-docs section, and include it when enabled

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://da4xlnrjyzf78.cloudfront.net/
cloudfront_id=E1LERG3RPFXU1K

File: config/feature.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (gibbardsteve) [reply config/feature.json]: Is it a big change to extract feature_disabled_navigation from this file, make that the default (wherever it used to live) and only use the feature file for settings associated with the feature?  I think you can then rename feature_active_navigation to navigation_settings.  This way the config in the feature file is only associated with features that are enabled.

File: sml_builder/__init__.py [hunk 13]
```diff
@@ -26,7 +27,10 @@
 
 @app.route("/")
 def index():
-    return render_template("index.html")
+    return render_template(
+        "index.html",
+        feature_1_enabled=sml_builder.utils.get_feature_config("feature_1"),
```
Comment (gibbardsteve) [inline sml_builder/__init__.py:32]: This probably needs removing as it was just Joe's example code?

File: sml_builder/__init__.py [hunk 14]
```diff
@@ -40,3 +44,24 @@ def string_to_paragraph(value):
...
+
+
+# Method provides a dictionary to the jinja templates, allowing variables
+# inside the dictionary to be directly accessed within the template files
+@app.context_processor
+def set_variables():
+    mkdocs = sml_builder.utils.get_feature_config("docs_integration")
+    navigation = {"navigation": {}}
+    nav_version = (
+        "feature_disabled_navigation"
```
Comment (gibbardsteve) [inline sml_builder/__init__.py:56]: I think here is where I'd prefer to just see:
if mkdocs["enabled"] is False
  Original nav code inline
else
  navigation_settings

File: sml_builder/__init__.py [hunk 15]
```diff
@@ -40,3 +44,20 @@ def string_to_paragraph(value):
     """Wraps passed string in <p> tags and converts newlines to <p> pairs"""
     body = escape(value).replace("\n\n", Markup("</p><p>"))
     return Markup(f"<p>{body}</p>")
+
+
+@app.context_processor
+def set_variables():
+    mkdocs = sml_builder.utils.get_feature_config("MKDOCS")
+    navigation = {"navigation": {}}
+    nav_version = (
```
[the remark this answers is not in the document]
Comment (gibbardsteve) [reply sml_builder/__init__.py:55]: It wasn't so much nav_version that was confusing me as the feature_active_navigate (or whatever it was before). I have made a suggestion re navigation_settings.

File: sml_builder/api_reference.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gibbardsteve) [reply sml_builder/api_reference.py]: Leave it for now. Hopefully it won't be long before we enable the routes.

File: sml_builder/__init__.py [hunk 16]
```diff
@@ -40,3 +44,24 @@ def string_to_paragraph(value):
     """Wraps passed string in <p> tags and converts newlines to <p> pairs"""
     body = escape(value).replace("\n\n", Markup("</p><p>"))
     return Markup(f"<p>{body}</p>")
+
+
+# Method provides a dictionary to the jinja templates, allowing variables
+# inside the dictionary to be directly accessed within the template files
+@app.context_processor
+def set_variables():
+    mkdocs = sml_builder.utils.get_feature_config("docs_integration")
```
Comment (gibbardsteve) [inline sml_builder/__init__.py:53]: Can we blanket replace mkdocs with docs_integration please.

**Review by gibbardsteve — COMMENTED** [review]
Thanks for those changes, some local vars are called mkdocs - if we can chnge those out that'd be great. Also a comment on the feature file.

File: sml_builder/__init__.py (hunk 16 above)
Comment (JasonBellONS) [reply sml_builder/__init__.py:53]: done

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://da4xlnrjyzf78.cloudfront.net/
cloudfront_id=E1LERG3RPFXU1K

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://da4xlnrjyzf78.cloudfront.net/
cloudfront_id=E1LERG3RPFXU1K

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://da4xlnrjyzf78.cloudfront.net/
cloudfront_id=E1LERG3RPFXU1K

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://da4xlnrjyzf78.cloudfront.net/
cloudfront_id=E1LERG3RPFXU1K

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://da4xlnrjyzf78.cloudfront.net/
cloudfront_id=E1LERG3RPFXU1K

**Comment by ons-spp-machine-user** [comment]
DEPLOY_URL=https://da4xlnrjyzf78.cloudfront.net/
cloudfront_id=E1LERG3RPFXU1K

**Review by gibbardsteve — APPROVED** [review]
Looks good

**Pull request merged**